From 6acacae983b860a1d95117845482e210b2285ce2 Mon Sep 17 00:00:00 2001
From: Ryan Coe <bluemrp9@gmail.com>
Date: Sat, 29 Feb 2020 15:51:56 -0800
Subject: [PATCH 2/3] silence shutdown messages

Signed-off-by: Ryan Coe <bluemrp9@gmail.com>
---
 init/init.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/init/init.c b/init/init.c
index efab5dcb44c01387a2ee30e0a98a085c17ebdfba..486bc2054039095d05255e0209ab06210d35fc87 100644
--- a/init/init.c
+++ b/init/init.c
@@ -762,16 +762,16 @@ static void run_shutdown_and_kill_processes(void)
 	 * shut things down gracefully... */
 	run_actions(SHUTDOWN);
 
-	message(L_CONSOLE | L_LOG, "The system is going down NOW!");
+	/*message(L_CONSOLE | L_LOG, "The system is going down NOW!");*/
 
 	/* Send signals to every process _except_ pid 1 */
 	kill(-1, SIGTERM);
-	message(L_CONSOLE, "Sent SIG%s to all processes", "TERM");
+	/*message(L_CONSOLE, "Sent SIG%s to all processes", "TERM");*/
 	sync();
 	sleep1();
 
 	kill(-1, SIGKILL);
-	message(L_CONSOLE, "Sent SIG%s to all processes", "KILL");
+	/*message(L_CONSOLE, "Sent SIG%s to all processes", "KILL");*/
 	sync();
 	/*sleep1(); - callers take care about making a pause */
 }
@@ -832,7 +832,7 @@ static void halt_reboot_pwoff(int sig)
 		m = "poweroff";
 		rb = RB_POWER_OFF;
 	}
-	message(L_CONSOLE, "Requesting system %s", m);
+	/*message(L_CONSOLE, "Requesting system %s", m);*/
 	pause_and_low_level_reboot(rb);
 	/* not reached */
 }
-- 
2.25.1

