From 0287244132036e311739f033c47569c933c35208 Mon Sep 17 00:00:00 2001
From: Ryan Coe <bluemrp9@gmail.com>
Date: Sat, 29 Feb 2020 15:51:56 -0800
Subject: [PATCH 1/1] silence shutdown messages

Signed-off-by: Ryan Coe <bluemrp9@gmail.com>
---
 init/init.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/init/init.c b/init/init.c
index 87086b48bc2118dceb3a434f69a844e1f03e5457..c6618c9c3eeb5db3950bfb47bf778402e28ab29c 100644
--- a/init/init.c
+++ b/init/init.c
@@ -768,16 +768,16 @@ static void run_shutdown_and_kill_processes(void)
 	 * shut things down gracefully... */
 	run_actions(SHUTDOWN);
 
-	message(L_CONSOLE | L_LOG, "The system is going down NOW!");
+	/*message(L_CONSOLE | L_LOG, "The system is going down NOW!");*/
 
 	/* Send signals to every process _except_ pid 1 */
 	kill(-1, SIGTERM);
-	message(L_CONSOLE, "Sent SIG%s to all processes", "TERM");
+	/*message(L_CONSOLE, "Sent SIG%s to all processes", "TERM");*/
 	sync();
 	sleep(1);
 
 	kill(-1, SIGKILL);
-	message(L_CONSOLE, "Sent SIG%s to all processes", "KILL");
+	/*message(L_CONSOLE, "Sent SIG%s to all processes", "KILL");*/
 	sync();
 	/*sleep(1); - callers take care about making a pause */
 }
@@ -844,7 +844,7 @@ static void halt_reboot_pwoff(int sig)
 		m = "poweroff";
 		rb = RB_POWER_OFF;
 	}
-	message(L_CONSOLE, "Requesting system %s", m);
+	/*message(L_CONSOLE, "Requesting system %s", m);*/
 	pause_and_low_level_reboot(rb);
 	/* not reached */
 }
-- 
2.25.1

