From b6ab61555ae226f44c7822edb6afab676f02193b Mon Sep 17 00:00:00 2001
From: Ryan Coe <bluemrp9@gmail.com>
Date: Sat, 14 May 2022 16:06:59 -0700
Subject: [PATCH 3/3] update init for rcinit usage

Signed-off-by: Ryan Coe <bluemrp9@gmail.com>
---
 init/init.c | 72 ++++++++++++++++++++++-------------------------------
 1 file changed, 30 insertions(+), 42 deletions(-)

diff --git a/init/init.c b/init/init.c
index efab5dcb44c01387a2ee30e0a98a085c17ebdfba..3d9ed1c1428a3fa0315f27a1283ac413c1dc3bf0 100644
--- a/init/init.c
+++ b/init/init.c
@@ -152,19 +152,22 @@
 #define CONSOLE_NAME_SIZE 32
 
 /* Default sysinit script. */
-#ifndef INIT_SCRIPT
-# define INIT_SCRIPT  "/etc/init.d/rcS"
+#ifndef STARTUP_SCRIPT
+# define STARTUP_SCRIPT  "/etc/init.d/rcS"
+#endif
+
+/* Default shutdown script. */
+#ifndef SHUTDOWN_SCRIPT
+# define SHUTDOWN_SCRIPT  "/etc/init.d/rcK"
 #endif
 
 /* Each type of actions can appear many times. They will be
  * handled in order. RESTART is an exception, only 1st is used.
  */
-/* Start these actions first and wait for completion */
-#define SYSINIT     0x01
 /* Start these after SYSINIT and wait for completion */
-#define WAIT        0x02
+#define WAIT        0x01
 /* Start these after WAIT and *dont* wait for completion */
-#define ONCE        0x04
+#define ONCE        0x02
 /*
  * NB: while SYSINIT/WAIT/ONCE are being processed,
  * SIGHUP ("reread /etc/inittab") will be processed only after
@@ -172,9 +175,11 @@
  * it will not be run, since init is already "past SYSINIT stage".
  */
 /* Start these after ONCE are started, restart on exit */
-#define RESPAWN     0x08
+#define RESPAWN     0x04
 /* Like RESPAWN, but wait for <Enter> to be pressed on tty */
-#define ASKFIRST    0x10
+#define ASKFIRST    0x08
+/* Start these actions first and wait for completion */
+#define SYSINIT     0x10
 /*
  * Start these on SIGINT, and wait for completion.
  * Then go back to respawning RESPAWN and ASKFIRST actions.
@@ -662,8 +667,7 @@ static void new_init_action(uint8_t action_type, const char *command, const char
 
 /* NOTE that if CONFIG_FEATURE_USE_INITTAB is NOT defined,
  * then parse_inittab() simply adds in some default
- * actions (i.e., runs INIT_SCRIPT and then starts a pair
- * of "askfirst" shells).  If CONFIG_FEATURE_USE_INITTAB
+ * actions.  If CONFIG_FEATURE_USE_INITTAB
  * _is_ defined, but /etc/inittab is missing, this
  * results in the same set of default behaviors.
  */
@@ -672,27 +676,28 @@ static void parse_inittab(void)
 #if ENABLE_FEATURE_USE_INITTAB
 	char *token[4];
 	parser_t *parser = config_open2("/etc/inittab", fopen_for_read);
+#endif
+
+	/* Startup scripts */
+	new_init_action(SYSINIT, STARTUP_SCRIPT, "");
+	/* Reboot on Ctrl-Alt-Del */
+	new_init_action(CTRLALTDEL, "reboot", "");
+	/* Shutdown scripts */
+	new_init_action(SHUTDOWN, SHUTDOWN_SCRIPT, "");
+	/* Restart init when a QUIT is received */
+	new_init_action(RESTART, "init", "");
 
+#if ENABLE_FEATURE_USE_INITTAB
 	if (parser == NULL)
 #endif
 	{
 		/* No inittab file - set up some default behavior */
-		/* Sysinit */
-		new_init_action(SYSINIT, INIT_SCRIPT, "");
 		/* Askfirst shell on tty1-4 */
 		new_init_action(ASKFIRST, bb_default_login_shell, "");
 //TODO: VC_1 instead of ""? "" is console -> ctty problems -> angry users
 		new_init_action(ASKFIRST, bb_default_login_shell, VC_2);
 		new_init_action(ASKFIRST, bb_default_login_shell, VC_3);
 		new_init_action(ASKFIRST, bb_default_login_shell, VC_4);
-		/* Reboot on Ctrl-Alt-Del */
-		new_init_action(CTRLALTDEL, "reboot", "");
-		/* Umount all filesystems on halt/reboot */
-		new_init_action(SHUTDOWN, "umount -a -r", "");
-		/* Swapoff on halt/reboot */
-		new_init_action(SHUTDOWN, "swapoff -a", "");
-		/* Restart init when a QUIT is received */
-		new_init_action(RESTART, "init", "");
 		return;
 	}
 
@@ -704,8 +709,7 @@ static void parse_inittab(void)
 				PARSE_NORMAL & ~(PARSE_TRIM | PARSE_COLLAPSE))) {
 		/* order must correspond to SYSINIT..RESTART constants */
 		static const char actions[] ALIGN1 =
-			"sysinit\0""wait\0""once\0""respawn\0""askfirst\0"
-			"ctrlaltdel\0""shutdown\0""restart\0";
+			"wait\0""once\0""respawn\0""askfirst\0";
 		int action;
 		char *tty = token[0];
 
@@ -760,19 +764,9 @@ static void run_shutdown_and_kill_processes(void)
 	/* Run everything to be run at "shutdown".  This is done _prior_
 	 * to killing everything, in case people wish to use scripts to
 	 * shut things down gracefully... */
-	run_actions(SHUTDOWN);
-
 	message(L_CONSOLE | L_LOG, "The system is going down NOW!");
-
-	/* Send signals to every process _except_ pid 1 */
-	kill(-1, SIGTERM);
-	message(L_CONSOLE, "Sent SIG%s to all processes", "TERM");
-	sync();
-	sleep1();
-
-	kill(-1, SIGKILL);
-	message(L_CONSOLE, "Sent SIG%s to all processes", "KILL");
-	sync();
+	run_actions(SHUTDOWN);
+	message(L_CONSOLE, "Reached target shutdown");
 	/*sleep1(); - callers take care about making a pause */
 }
 
@@ -810,7 +804,6 @@ static void run_shutdown_and_kill_processes(void)
 static void halt_reboot_pwoff(int sig) NORETURN;
 static void halt_reboot_pwoff(int sig)
 {
-	const char *m;
 	unsigned rb;
 
 	/* We may call run() and it unmasks signals,
@@ -823,16 +816,12 @@ static void halt_reboot_pwoff(int sig)
 
 	run_shutdown_and_kill_processes();
 
-	m = "halt";
 	rb = RB_HALT_SYSTEM;
 	if (sig == SIGTERM) {
-		m = "reboot";
 		rb = RB_AUTOBOOT;
 	} else if (sig == SIGUSR2) {
-		m = "poweroff";
 		rb = RB_POWER_OFF;
 	}
-	message(L_CONSOLE, "Requesting system %s", m);
 	pause_and_low_level_reboot(rb);
 	/* not reached */
 }
@@ -1106,7 +1095,7 @@ int init_main(int argc UNUSED_PARAM, char **argv)
 
 	/* Make sure environs is set to something sane */
 	putenv((char *) "HOME=/");
-	putenv((char *) bb_PATH_root_path);
+	putenv((char *) "PATH=/usr/sbin:/usr/bin");
 	putenv((char *) "SHELL=/bin/sh");
 	putenv((char *) "USER=root"); /* needed? why? */
 
@@ -1130,8 +1119,7 @@ int init_main(int argc UNUSED_PARAM, char **argv)
 
 		/* NOTE that if CONFIG_FEATURE_USE_INITTAB is NOT defined,
 		 * then parse_inittab() simply adds in some default
-		 * actions (i.e., INIT_SCRIPT and a pair
-		 * of "askfirst" shells) */
+		 * actions */
 		parse_inittab();
 	}
 
-- 
2.25.1

